Add `Beatmapping/Flow` (#7407)
ppy/osu-wiki: wiki/Beatmapping/Flow/en.md

---
tags:
  - linear flow
  - oval flow
  - circular flow
---

# Flow

> **Flow** is a mapping concept that describes natural and intuitive movement patterns of a player during gameplay. Flow, or the lack thereof, is almost always related to the [osu! game mode](/wiki/Game_mode/osu!) due to the two-dimensional playfield. By utilising the vast variety of [patterns](/wiki/Beatmap/Pattern) and slider shapes possible in osu! beatmaps, mappers are able to direct a player's cursor around the playfield in various ways, yielding different effects on the gameplay experience.

Reviewer: > Flow, or the lack thereof, is almost always related to the [osu! game mode](/wiki/Game_mode/osu!) due to the two-dimensional playfield

This is not true as it also applies to the gamemode osu!catch

Author: I left a note for someone (maybe it could be the future you?) to pick up

If a mapper deliberately ignores the concepts of flow, or uses their opposites, they utilise [antiflow](/wiki/Beatmapping/Antiflow).

## Key components

> While flow is mostly subjective, as movement preferences differ from one person to another, there are a few ideas that give the feeling and impression of flow on a majority of patterns or whole beatmaps.

Reviewer 2: i feel like there should be more emphasis on the subjectivity of this somehow, since lots of mappers think flow is an outdated term that should never be used nowadays lol

### Smooth movement

> The cursor movement required to pass the map is stable most of the time and doesn't contain sharp, abrupt turns, or fast change of pace. Often, a player's cursor follows a smooth curve that spans for several patterns. The curve contains lots of circular or semicircular segments (so-called *oval or circular flow*) going in different directions, mixed with "neutral", non-demanding patterns. Such patterns don't contribute to the general direction, or only change it slightly to ease the transition to the next pattern.

Reviewer 2: might be worth mentioning that circular flow for too long in the same direction is viewed as uncomfortable because mousedrift

> The cursor movement required to pass the map is stable most of the time and doesn't contain sharp, abrupt turns, or fast change of pace. Often, a player's cursor follows a smooth curve that spans for several patterns. The curve contains lots of circular or semicircular segments (so-called *oval or circular flow*) going in different directions, mixed with "neutral", non-demanding patterns. Such patterns don't contribute to the general direction, or only change it slightly to ease the transition to the next pattern.

Reviewer 3: ```suggestion
The cursor movement required to pass the map is stable most of the time and doesn't contain sharp, abrupt turns, or fast change of pace. Often, a player's cursor follows a smooth curve that spans for several patterns. The curve contains lots of circular or semicircular segments (so-called *oval* or *circular flow*) going in different directions, mixed with "neutral", non-demanding patterns. Such patterns don't contribute to the general direction, or only change it slightly to ease the transition to the next pattern.
```

Linear movement may also be considered another form of flow, albeit less common, since it doesn't enforce any specific cursor trajectory type.

### Repetition

Another kind of movement that feels natural is repetition. By clearing similar patterns that only differ slightly, such as back-and-forth [jumps](/wiki/Beatmap/Pattern/Jump) or a series of slider pairs, a player gets more comfortable with the map. This happens because every next movement follows an already familiar path and becomes easy to predict.

### Visual implications

To further make the beatmap look and feel more natural, mappers can use [streams](/wiki/Beatmap/Pattern/Stream) or sliders with the curves that follow the suggested movement direction. Since sliders' bodies may not be followed closely, as opposed to [slider ticks](/wiki/Beatmapping/Slider_tick), they may be adjusted and bent in lots of ways. This fact may be used both in favour of an already present movement trajectory, and to help the player pick up a new movement type section.

## See also

- [Analyzing Flow in Beatmaps](https://osu.ppy.sh/community/forums/topics/86329), a guide by ![][flag_US] [Charles445](https://osu.ppy.sh/users/85000)

> - [Mapping Analysis #1 - Flow](https://osu.ppy.sh/home/news/2013-09-15-mapping-analysis-1-flow), a newspost by ![][flag_FR] [Shiro](https://osu.ppy.sh/users/113005)

Reviewer 3: ```suggestion
- [Mapping Analysis #1 - Flow](https://osu.ppy.sh/home/news/2013-09-15-mapping-analysis-1-flow), a news post by ![][flag_FR] [Shiro](https://osu.ppy.sh/users/113005)
```

> - *[flow + circles](https://www.youtube.com/watch?v=ig7nTSbWwVc)* and *[flow + sliders](https://www.youtube.com/watch?v=GYGatwQszwQ)*, the episodes of the [osu!mapping](/wiki/Community/Video_series/osu!mapping) YouTube series

Reviewer 3: ```suggestion
- *[flow + circles](https://www.youtube.com/watch?v=ig7nTSbWwVc)* and *[flow + sliders](https://www.youtube.com/watch?v=GYGatwQszwQ)*, episodes of the [osu!mapping](/wiki/Community/Video_series/osu!mapping) YouTube series by ![][flag_US] [pishifat](https://osu.ppy.sh/users/3178418)
```

---
tags:
  - linear flow
  - oval flow
  - circular flow
---

# Flow

**Flow** is a mapping concept that describes natural and intuitive movement patterns of a player's cursor during gameplay. Flow, or the lack thereof, is related to the [osu! game mode](/wiki/Game_mode/osu!) due to its two-dimensional playfield. By utilising the vast variety of [patterns](/wiki/Beatmap/Pattern) and slider shapes possible in osu! beatmaps, mappers are able to direct a player's cursor around the playfield in various ways, yielding different effects on the gameplay experience.

<!-- TODO: this applies to osu!catch as well, the article needs expansion: https://github.com/ppy/osu-wiki/pull/7407#pullrequestreview-988657741 -->

> Flow is considered a **subjective** and often opinionated matter due to how different mappers and players design or perceive movement. Some aspects of flow may impact gameplay in different degrees depending on [input device](/wiki/Gameplay/Input_device), such as [graphics tablet](/wiki/Gameplay/Input_device/Graphics_tablet) or [mouse](/wiki/Gameplay/Input_device/Mouse), or the player's acquired skill or experience.

Reviewer 2: ```suggestion
Flow is considered a **subjective** and often opinionated matter due to how different mappers and players design or perceive movement. Some aspects of flow may impact gameplay to different degrees depending on the player's [input device](/wiki/Gameplay/Input_device), such as [graphics tablet](/wiki/Gameplay/Input_device/Graphics_tablet) or [mouse](/wiki/Gameplay/Input_device/Mouse), or the player's acquired skill or experience.
```

The opposite of flow is called [antiflow](/wiki/Beatmapping/Antiflow), which conversely involves unintuitive or unnatural movement patterns.

## Key components

While movement preferences vary from one person to another, there are a few ideas that give the feeling and impression of flow on a majority of patterns or whole beatmaps.

### Smooth movement

Series of hit objects that require smooth movement, without sharp, abrupt turns, or fast changes of pace, can be considered comfortable to play. In such patterns, the player's cursor may follow a smooth curve throughout. These curves are usually characterised by circular or semicircular segments (so-called *oval* or *circular flow*) going either clockwise or anticlockwise. Beatmaps often switch the direction of the circular flow at regular intervals,[^circular-flow] where the transitions can often accompanied by "neutral" patterns to ease the transition to the next pattern.

Linear movement may also be considered another form of smooth flow, albeit less common, since it does not enforce any specific cursor trajectory type.

### Repetition

> Another form of natural movement is repetition. Patterns with similar repeated movements that only differ slightly, such as back-and-forth [jumps](/wiki/Beatmap/Pattern/Jump) or a series of slider pairs, are comfortable to play due to subsequent movements following already familiar paths, making them easy to predict. Repeated patterns may become more and more comfortable throughout a map for the same reason.

Reviewer 3: i think this was trying to describe two things at the same time, with the "more comfortable with the map" thing. that said i'm not sure if the last sentence here is relevant to flow, although it may be true

Reviewer 2: last sentence is valid

### Visual implications

> To further make a beatmap look and feel more natural, mappers may use [streams](/wiki/Beatmap/Pattern/Stream) or sliders with curves that reinforce suggested direction of movement. Since sliders bodies do not need to be followed perfectly, because the cursor only needs to reach the [slider ticks](/wiki/Beatmapping/Slider_tick), they may be adjusted and bent in lots of ways without affecting the movement required to clear them. This fact may be used both in favour of an already present movement trajectory, or to help the player pick up a new movement type.

Reviewer 3: if this is about slider leniency and the freedom in slider shapes it allows, then what do steams have to do with it? sure stream flow is a thing, but the paragraph continued on to only talk about something specific to sliders

if there's something to say on the visual aspect of streams, then perhaps it can receive its own paragraph

also not entirely sure what the last sentence is about

```suggestion
To further make a beatmap look and feel more natural, mappers may use sliders with different shapes that reinforce suggested direction of movement. Since sliders bodies do not need to be followed perfectly, because the cursor only needs to reach the [slider ticks](/wiki/Beatmapping/Slider_tick), they may be adjusted and bent in lots of ways without affecting the movement required to clear them. This fact may be used both in favour of an already present movement trajectory, or to help the player pick up a new movement type.
```

Reviewer 2: agree w the streams removal
the last sentence is relevant imo. might be better to word as 
`This allows slidershapes to both complement an existing movement trajectory, or help the player pick up a new movement trajectory.`
